### PR TITLE
Query: Convert IS operator in predicate to comparison with discriminator

### LINKED ISSUE #5007: Query: Inheritance filter with "is" performs client evaluation
The following query performs client filtering:
```cs
        [Fact]
        public virtual void Can_use_is_kiwi()
        {
            using (var context = CreateContext())
            {
                var kiwis = context.Set<Animal>().Where(a => a is Kiwi).ToList();

                Assert.Equal(1, kiwis.Count);
            }
        }
```

With generated SQL:
```sql
SELECT [a].[Species], [a].[CountryId], [a].[Discriminator], [a].[Name], [a].[EagleId], [a].[IsFlightless], [a].[Group], [a].[FoundOn]
FROM [Animal] AS [a]
WHERE [a].[Discriminator] IN (N'Kiwi', N'Eagle')
```

Note, the same query with the where clause replaced with `OfType<Kiwi>()` filters on the server:
```sql
SELECT [a].[Species], [a].[CountryId], [a].[Discriminator], [a].[Name], [a].[EagleId], [a].[IsFlightless], [a].[Group], [a].[FoundOn]
FROM [Animal] AS [a]
WHERE [a].[Discriminator] = N'Kiwi'
```

### PR DESCRIPTION
**Pull request by smitpatel**
Resolves #5007

### CODE REVIEW

File: src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs [hunk 1]
```diff
@@ -830,6 +835,68 @@ protected virtual void CheckClientEval([NotNull] object expression)
...
+                }
+
+                return Expression.Constant(true, typeof(bool));
+            }
+        }
+
+        private static readonly MethodInfo _efPropertyMethod
+            = typeof(EF).GetTypeInfo().GetDeclaredMethod(nameof(EF.Property));
+
+        private static Expression CreatePropertyExpression(Expression target, IProperty property)
```
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs]: these methods are being used in 2 places (other one being `EntityEqualityRewritingExpressionVisitor`), any good place to refactor?

File: src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs [hunk 2]
```diff
@@ -701,6 +703,9 @@ public override void VisitWhereClause(WhereClause whereClause, QueryModel queryM
 
             if (!requiresClientFilter)
             {
+                var discriminatorPredicate = new PedicateTypeIsExpressionTranslatingVisitor(this).Visit(whereClause.Predicate);
```
Comment (maumar) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs]: typo

File: src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs (hunk 1 above)
Comment (maumar) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs]: Maybe EntityQueryModelVisitor? we already have a IsPropertyMethod on there

File: test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/InheritanceSqlServerTest.cs [hunk 3]
```diff
@@ -30,7 +30,18 @@ public override void Can_use_is_kiwi()
             Assert.Equal(
                 @"SELECT [a].[Species], [a].[CountryId], [a].[Discriminator], [a].[Name], [a].[EagleId], [a].[IsFlightless], [a].[Group], [a].[FoundOn]
 FROM [Animal] AS [a]
-WHERE [a].[Discriminator] IN (N'Kiwi', N'Eagle')",
+WHERE [a].[Discriminator] IN (N'Kiwi', N'Eagle') AND ([a].[Discriminator] = N'Kiwi')",
```
Comment (maumar) [inline test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/InheritanceSqlServerTest.cs]: that's something we could probably optimize (in the future)

**Comment by maumar** [comment]
:shipit:

File: src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs [hunk 4]
```diff
@@ -830,6 +835,68 @@ protected virtual void CheckClientEval([NotNull] object expression)
             }
         }
 
+        private class PedicateTypeIsExpressionTranslatingVisitor : ExpressionVisitorBase
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs]: *Predicate*

File: src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs (hunk 1 above)
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs]: Even introducing a new helper class would be preferable here.

File: src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs (hunk 4 above)
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs]: I don't think this should just be for predicates. Add some tests with *is* in the projection etc.

File: test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/InheritanceSqlServerTest.cs (hunk 3 above)
Comment (smitpatel) [inline test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/InheritanceSqlServerTest.cs]: filed #5230

**Comment by smitpatel** [comment]
Updated with feedback.
The translatingvisitor now works on whole query model so projection/predicate both are translated. (added test for both).
Also removed generation of `Where 1 = 1` in query. If predicate is true then we don't generate where clause.

File: src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs [hunk 5]
```diff
@@ -830,6 +835,58 @@ protected virtual void CheckClientEval([NotNull] object expression)
             }
         }
 
+        private class TypeIsExpressionTranslatingVisitor : ExpressionVisitorBase
+        {
+            private readonly RelationalQueryModelVisitor _queryModelVisitor;
+
+            public TypeIsExpressionTranslatingVisitor(RelationalQueryModelVisitor relationalQueryModelVisitor)
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs]: Minor but consider reducing coupling here by only passing the required dependencies: Model and RelationalAnnotationProvider. Then you can revert the RelationalAnnotationProvider property change, too.

File: src/Microsoft.EntityFrameworkCore/Query/ExpressionVisitors/Internal/EntityEqualityRewritingExpressionVisitor.cs [hunk 6]
```diff
@@ -81,27 +81,17 @@ var newRightExpression
...
                     Expression.NewArrayInit(
                         typeof(object),
                         properties
-                            .Select(p => Expression.Convert(CreatePropertyExpression(target, p), typeof(object)))
+                            .Select(p => Expression.Convert(EntityQueryModelVisitor.CreatePropertyExpression(target, p), typeof(object)))
                             .Cast<Expression>()
                             .ToArray()));
         }
 
-        private static readonly MethodInfo _efPropertyMethod
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore/Query/ExpressionVisitors/Internal/EntityEqualityRewritingExpressionVisitor.cs]: Nice.

**Comment by smitpatel** [comment]
updated.

**Comment by anpete** [comment]
:shipit:

**Pull request merged**